### PR TITLE
[Federation][kubefed] Create a dedicated service account for federation controller manager in the host cluster and give it appropriate permissions.

### PR DESCRIPTION
**Pull request by madhusudancs**
Ref: Issue #39555

cc @kubernetes/sig-federation-pr-reviews @kubernetes/sig-auth-misc @kubernetes/sig-auth-pr-reviews 

```release-note
kubefed init creates a service account for federation controller manager in the federation-system namespace and binds that service account to the federation-system:federation-controller-manager role that has read and list access on secrets in the federation-system namespace. 
```

### CODE REVIEW

File: federation/pkg/kubefed/init/init.go [hunk 1]
```diff
@@ -520,7 +545,46 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: roleName,
+		},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("get", "list", "watch").Groups(legacyAPIGroup).Resources("secrets").RuleOrDie(),
+		},
+	}
+	rolebinding := rbac.NewRoleBinding(roleName, namespace).SAs(namespace, saName).BindingOrDie()
+
+	if dryRun {
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init.go]: Did you mean if !dryRun?

File: federation/pkg/kubefed/init/init.go [hunk 2]
```diff
@@ -62,6 +63,16 @@ const (
 	AdminCN                     = "admin"
 	HostClusterLocalDNSZoneName = "cluster.local."
 
+	// User name used by federation controller manager to make
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init.go]: Dont we need similar constant for federation controller manager to talk to host cluster?

File: federation/pkg/kubefed/init/init.go [hunk 3]
```diff
@@ -520,7 +545,46 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+		},
+	}
+	if dryRun {
+		return sa, nil
+	}
+	return clientset.Core().ServiceAccounts(namespace).Create(sa)
+}
+
+func createRoleBindings(clientset *client.Clientset, namespace, saName string, dryRun bool) (*rbac.Role, *rbac.RoleBinding, error) {
+	roleName := "federation-system:federation-controller-manager"
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init.go]: Use the constants here?

**Comment by nikhiljindal** [comment]
Would be great to have a unit test as well :)

File: federation/pkg/kubefed/init/init.go [hunk 4]
```diff
@@ -520,7 +545,46 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+	role := &rbac.Role{
+		// a role to use for bootstrapping the federation-controller-manager so it can access
```
Comment (liggitt) [inline federation/pkg/kubefed/init/init.go]: just to clarify, this is for watching secrets stored in the federation server, where it gets credentials for other servers?

File: federation/pkg/kubefed/init/init.go [hunk 5]
```diff
@@ -520,7 +545,46 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+	}
+	if dryRun {
+		return sa, nil
+	}
+	return clientset.Core().ServiceAccounts(namespace).Create(sa)
+}
+
+func createRoleBindings(clientset *client.Clientset, namespace, saName string, dryRun bool) (*rbac.Role, *rbac.RoleBinding, error) {
+	roleName := "federation-system:federation-controller-manager"
+	role := &rbac.Role{
```
Comment (liggitt) [inline federation/pkg/kubefed/init/init.go]: is the intent to have these permissions within a single namespace or across all namespaces? to grant this across all namespaces, this needs to be a clusterrole

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal @liggitt 

> Did you mean if !dryRun?

Oops, you are right. Fixed it.

> Would be great to have a unit test as well :)

Added some unit tests. I did not really find a framework for testing the RBACs in unit tests. I will continue investigating that and will add a few tests in a future PR.

File: federation/pkg/kubefed/init/init.go [hunk 6]
```diff
@@ -62,6 +63,16 @@ const (
 	AdminCN                     = "admin"
 	HostClusterLocalDNSZoneName = "cluster.local."
 
+	// User name used by federation controller manager to make
+	// calls to federation API server.
+	ControllerManagerUser = "federation-controller-manager"
+
+	// Name of the ServiceAccount used by the federation controller manager.
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init.go]: Please clarify in the comment is the SA used to talk to federation apiserver or to talk to registered clusters or just the host cluster or all of them :)

File: federation/pkg/kubefed/init/init.go [hunk 7]
```diff
@@ -220,7 +231,21 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 	}
 
 	// 7. Create federation controller manager
-	_, err = createControllerManager(hostClientset, initFlags.FederationSystemNamespace, initFlags.Name, svc.Name, cmName, image, cmKubeconfigName, dnsZoneName, dnsProvider, dryRun)
+	// 7a. Create service accounts for federation controller manager
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init.go]: "Create a service account in the host cluster for federation controller manager"

**Comment by nikhiljindal** [comment]
thanks, lgtm with 2 minor comments on comments :)

File: federation/pkg/kubefed/init/init.go (hunk 6 above)
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Done.

File: federation/pkg/kubefed/init/init.go (hunk 7 above)
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Done.

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal done. PTAL.

**Comment by madhusudancs** [comment]
@k8s-bot unit test this

File: pkg/apis/rbac/helpers.go [hunk 8]
```diff
@@ -254,3 +254,63 @@ func (r *ClusterRoleBindingBuilder) Binding() (ClusterRoleBinding, error) {
...
 }
+
+// +k8s:deepcopy-gen=false
+// RoleBindingBuilder let's us attach methods. It is similar to
+// ClusterRoleBindingBuilder above.
+type RoleBindingBuilder struct {
+	RoleBinding RoleBinding
+}
+
+func NewRoleBinding(roleName, namespace string) *RoleBindingBuilder {
```
Comment (lavalamp) [inline pkg/apis/rbac/helpers.go]: Document that at least one of the functions below must be called.

File: federation/pkg/kubefed/init/init.go [hunk 9]
```diff
@@ -520,7 +547,53 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+			Namespace: namespace,
+			Labels:    componentLabel,
+		},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("get", "list", "watch").Groups(legacyAPIGroup).Resources("secrets").RuleOrDie(),
+		},
+	}
+
+	rolebinding := rbac.NewRoleBinding(roleName, namespace).SAs(namespace, saName).BindingOrDie()
+	rolebinding.Namespace = namespace
```
Comment (lavalamp) [inline federation/pkg/kubefed/init/init.go]: wasn't the namespace already set by the above?

File: federation/pkg/kubefed/init/init.go [hunk 10]
```diff
@@ -520,7 +547,53 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
...
+			Name:      roleName,
+			Namespace: namespace,
+			Labels:    componentLabel,
+		},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("get", "list", "watch").Groups(legacyAPIGroup).Resources("secrets").RuleOrDie(),
+		},
+	}
+
+	rolebinding := rbac.NewRoleBinding(roleName, namespace).SAs(namespace, saName).BindingOrDie()
```
Comment (lavalamp) [inline federation/pkg/kubefed/init/init.go]: XOrDie methods shouldn't be called in production code, IMO.

File: pkg/apis/rbac/helpers.go [hunk 11]
```diff
@@ -254,3 +254,63 @@ func (r *ClusterRoleBindingBuilder) Binding() (ClusterRoleBinding, error) {
...
+			RoleRef: RoleRef{
+				APIGroup: GroupName,
+				Kind:     "Role",
+				Name:     roleName,
+			},
+		},
+	}
+}
+
+func (r *RoleBindingBuilder) Groups(groups ...string) *RoleBindingBuilder {
```
Comment (lavalamp) [inline pkg/apis/rbac/helpers.go]: Public methods should have a godoc string.

File: pkg/apis/rbac/helpers.go [hunk 12]
```diff
@@ -254,3 +254,63 @@ func (r *ClusterRoleBindingBuilder) Binding() (ClusterRoleBinding, error) {
 
 	return r.ClusterRoleBinding, nil
 }
+
+// +k8s:deepcopy-gen=false
```
Comment (lavalamp) [inline pkg/apis/rbac/helpers.go]: I'm not a huge fan of having helpers in the internal types api packages. @deads2k 

I guess you're not making it worse, so I'm not asking for you to solve this problem.

File: federation/pkg/kubefed/init/init.go (hunk 10 above)
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Dealing with the error is not that bad in this case. But for example, not having `XOrDie` in the `Rules` slice above makes it really hard to use the builder pattern.

But changed this to handle the error.

File: federation/pkg/kubefed/init/init.go (hunk 9 above)
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Removed this. This was a left over from my manual tests.

File: pkg/apis/rbac/helpers.go (hunk 8 above)
Comment (madhusudancs) [inline pkg/apis/rbac/helpers.go]: Done.

File: pkg/apis/rbac/helpers.go (hunk 11 above)
Comment (madhusudancs) [inline pkg/apis/rbac/helpers.go]: Done.

**Review by madhusudancs — COMMENTED** [review]
@lavalamp addressed the comments. PTAL.

**Comment by madhusudancs** [comment]
@k8s-bot gce etcd3 e2e test this

File: federation/pkg/kubefed/init/init.go (hunk 3 above)
Comment (lavalamp) [inline federation/pkg/kubefed/init/init.go]: nit: if it's constant, declare it as a constant.

Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Will do that in a follow up. Thanks for the approval.

File: pkg/apis/rbac/helpers.go (hunk 12 above)
Comment (deads2k) [inline pkg/apis/rbac/helpers.go]: > I'm not a huge fan of having helpers in the internal types api packages. @deads2k

RBAC really demonstrates the issue with using external versions.  The RBAC authorizer doesn't care which version of the API you use, but you run into crazy problems with caches, informers, conversions, and helpers if you don't use the internal types.

I suspect that if you saw the code duplication involved in having v1beta1 and v1alpha1 support, you'd be displeased.

**Pull request merged**